Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,687 @@
+      case .clientOpenServerIdle(var clientOpenState):
+        let request = try clientOpenState.framer.next(compressor: clientOpenState.compressor)
+        self.state = .clientOpenServerIdle(clientOpenState)
+        return request
+      case .clientOpenServerOpen(var clientOpenState):
+        let request = try clientOpenState.framer.next(compressor: clientOpenState.compressor)
+        self.state = .clientOpenServerOpen(clientOpenState)
+        return request
+      case .clientOpenServerClosed:
+        // Nothing to do: no point in sending request if server is closed.
+        return nil
+      case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Can't send request if client is closed.")
+      }
+    }
+    
+    mutating func receive(metadata: Metadata, endStream: Bool) throws -> OnMetadataReceived {
+      // This is metadata received by the client from the server.
+      // It can be initial, which confirms that the server is now open;
+      // or an END_STREAM trailer, meaning the response is over.
+      if endStream {
+        try self.clientReceivedEndHeader()
+      } else {
+        try self.clientReceivedMetadata()
+      }
+      return .doNothing
+    }
+    
+    mutating func clientReceivedEndHeader() throws {
+      switch self.state {
+      case .clientIdleServerIdle:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client can't have received a stream end trailer if both client and server are idle.")
+      case .clientOpenServerIdle:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server cannot have sent an end stream header if it is still idle.")
+      case .clientOpenServerOpen(let state):
+        self.state = .clientOpenServerClosed(.init(previousState: state))
+      case .clientOpenServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server is already closed, can't have received the end stream trailer twice.")
+      case .clientClosedServerIdle:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server cannot have sent end stream trailer if it is idle.")
+      case .clientClosedServerOpen(let state):
+        state.compressor?.end()
+        state.decompressor?.end()
+        self.state = .clientClosedServerClosed
+      case .clientClosedServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server cannot have sent end stream trailer if it is already closed.")
+      }
+    }
+    
+    mutating func clientReceivedMetadata() throws {
+      switch self.state {
+      case .clientIdleServerIdle:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server cannot have sent metadata if the client is idle.")
+      case .clientOpenServerIdle(let state):
+        self.state = .clientOpenServerOpen(.init(previousState: state))
```

Comment:
We need to inspect the response headers because the server will indicate what compression algorithm the server is using, if any.

We also need to check the ":status" pseudo header (must be 200, although some http error codes will map directly to a gRPC status and in such a case we need to move to closed and forward an appropriate `RPCStatus` see also https://github.com/grpc/grpc/blob/master/doc/http-grpc-status-mapping.md) and content-type.

Reply:
This function will also need to deal with converting the `HPACKHeaders` to `Metadata`